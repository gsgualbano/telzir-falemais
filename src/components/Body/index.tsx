import React, { useState } from 'react';

import { Form } from '@rocketseat/unform';
import * as Yup from 'yup';

import { Animated } from 'react-animated-css';
import {
  Container,
  Box,
  SelectBox,
  Label,
  TextInput,
  Button,
  Text,
  BoxWrapper,
  TextWrapper,
  FormGroup,
} from './styles';
import BarChart from '../BarChart';

interface FormData {
  origem: string,
  destino: string,
  tempo: number,
  plano: string,
  precos: { comFaleMais: number, semFaleMais: number },
  economia: string,
}

// validação do formulário
const schema = Yup.object().shape({
  origem: Yup.string().required('DDD Origem é necessário'),
  destino: Yup.string().required('DDD Destino é necessário'),
  tempo: Yup.number()
    .typeError('Número inválido')
    .required('Necessário preencher o tempo de ligação'),
  plano: Yup.string().required('Selecione algum dos planos'),
});

/**
 * Corpo da página
 */
const Body = (props: {}) => {
  /** Hooks */
  const [showChart, setShowChart] = useState<boolean>(false); // mostrar o gráfico com os preços
  const [formData, setFormData] = useState<FormData>({
    origem: '',
    destino: '',
    tempo: 0,
    plano: '',
    precos: {
      comFaleMais: 0,
      semFaleMais: 0,
    },
    economia: '',
  });

  const locais = [
    { id: '11', title: '11' },
    { id: '16', title: '16' },
    { id: '17', title: '17' },
    { id: '18', title: '18' },
  ];

  const planos = [
    { id: 'FaleMais 30', title: 'FaleMais 30' },
    { id: 'FaleMais 60', title: 'FaleMais 60' },
    { id: 'FaleMais 120', title: 'FaleMais 120' },
  ];

  const calculaPrecos = (origem: string, destino: string, tempo: number, decrecimo: number) => {
    let comFaleMais = 0;
    let semFaleMais = 0;
    let tempoComFaleMais = tempo - decrecimo;

    // zera o tempo caso ele seja menor que o tempo limite do plano
    if (tempoComFaleMais < 0) {
      tempoComFaleMais = 0;
    }

    if (origem === '11' && destino === '16') {
      comFaleMais = 1.9 * tempoComFaleMais;
      semFaleMais = 1.9 * tempo;
    } else if (origem === '16' && destino === '11') {
      comFaleMais = 2.9 * tempoComFaleMais;
      semFaleMais = 2.9 * tempo;
    } else if (origem === '11' && destino === '17') {
      comFaleMais = 1.7 * tempoComFaleMais;
      semFaleMais = 1.7 * tempo;
    } else if (origem === '17' && destino === '11') {
      comFaleMais = 2.7 * tempoComFaleMais;
      semFaleMais = 2.7 * tempo;
    } else if (origem === '11' && destino === '18') {
      comFaleMais = 0.9 * tempoComFaleMais;
      semFaleMais = 0.9 * tempo;
    } else if (origem === '18' && destino === '11') {
      comFaleMais = 1.9 * tempoComFaleMais;
      semFaleMais = 1.9 * tempo;
    } else {
      comFaleMais = NaN;
      semFaleMais = NaN;
    }

    comFaleMais *= 1.1; // acrescimo de 10% da tarifa excedente

    return { comFaleMais, semFaleMais };
  };

  const handleSubmit = (data: any) => {
    const {
      origem,
      destino,
      tempo,
      plano,
    } = data;

    let precos = { comFaleMais: 0, semFaleMais: 0 };

    switch (plano) {
      case 'FaleMais 30':
        precos = calculaPrecos(origem, destino, tempo, 30); // FaleMais30
        break;
      case 'FaleMais 60':
        precos = calculaPrecos(origem, destino, tempo, 60); // FaleMais60
        break;
      case 'FaleMais 120':
        precos = calculaPrecos(origem, destino, tempo, 120); // FaleMais120
        break;
    }

    let economia;

    // se os dois valores não forem zero e nem NaN
    if (precos.comFaleMais && precos.semFaleMais) {
      const porcentagem = 100 - ((precos.comFaleMais / precos.semFaleMais) * 100);
      economia = `${porcentagem.toFixed(2)}%`;
    } else if (precos.comFaleMais === 0 && precos.semFaleMais !== 0) {
      economia = '100%';
    } else {
      economia = '-';
    }

    setFormData({
      origem,
      destino,
      tempo,
      plano,
      precos,
      economia,
    });

    setShowChart(!showChart);
  };

  const renderChart = (): JSX.Element => <BarChart formData={formData} isVisible={showChart} setShowChart={setShowChart} />;

  return (
    <Container>
      <TextWrapper>
        <Text>
          Com o nosso novo plano FaleMais, o cliente pode falar de graça até um determinado tempo e
          depois só paga os minutos excedentes. calcule e veja as vantagens
        </Text>
      </TextWrapper>
      <BoxWrapper>
        <Box>
          <Animated animationIn="fadeIn" animationOut="fadeOut" isVisible={!showChart}>
            {!showChart && (
              <Form schema={schema} onSubmit={handleSubmit}>
                <FormGroup>
                  <Label htmlFor="origem">DDD Origem</Label>
                  <SelectBox id="origem" name="origem" options={locais} />
                </FormGroup>

                <FormGroup>
                  <Label htmlFor="destino">DDD Destino</Label>
                  <SelectBox id="destino" name="destino" options={locais} />
                </FormGroup>

                <FormGroup>
                  <Label htmlFor="tempo">Tempo de ligação(em minutos)</Label>
                  <TextInput id="tempo" name="tempo" placeholder="" />
                </FormGroup>

                <FormGroup>
                  <Label htmlFor="plano">Plano</Label>
                  <SelectBox id="plano" name="plano" options={planos} />
                </FormGroup>
                <Button data-testid="form-button">Calcular</Button>
              </Form>
            )}
          </Animated>
          {showChart && renderChart()}
        </Box>
      </BoxWrapper>
    </Container>
  );
};

export default Body;
