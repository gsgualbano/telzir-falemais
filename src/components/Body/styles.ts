import styled from 'styled-components';
import { Select, Input } from '@rocketseat/unform';

export const Container = styled.div`
  flex: 1;
  display: flex;
  flex-direction: column;
`;

export const TextWrapper = styled.div`
  display: flex;
  justify-content: center;
`;

export const Text = styled.p`
  font-weight: 600;
  font-size: 16px;
  text-align: center;
  color: #fff;
  margin-bottom: 15px;
  padding: 5px;
  width: 100%;
  max-width: 520px;
`;

export const BoxWrapper = styled.div`
  display: flex;
  justify-content: center;
  
  @media(max-width: 768px) { /* Telas de IPAD ou menores */
   flex: 1; // usar o 100% do espaço disponível
  }
`;

export const Box = styled.div`
  display: flex;
  flex-direction: column;
  background: #fff;
  box-shadow: 0 15px 20px 0 rgba(0, 0, 0, 0.5);
  border-radius: 10px;
  padding: 20px 15px;
  width: 100%;
  max-width: 520px;
  
  @media(max-width: 768px) { /* Telas de IPAD ou menores */
    border-radius: 10px 10px 0 0; // borda somente nas pontas superiores
  }
`;

export const FormGroup = styled.div`
  display: flex;
  flex-direction: column;
  margin-bottom: 20px;

  & span {
    color: red;
  }
`;

export const SelectBox = styled(Select)`
  background: #ebebeb;
  border-radius: 5px;
  border: none;
  padding: 10px;
  outline: none;
`;

export const TextInput = styled(Input).attrs({
  type: 'number',
})`
  background: #ebebeb;
  border-radius: 5px;
  border: none;
  padding: 10px;
  outline: none;
`;

export const Label = styled.label`
  color: #8B8B8B;
  margin-bottom: 5px;
  font-weight: 600;
`;

export const Button = styled.button`
  margin-top: 10px;
  width: 100%;
  background: #3C40C6;
  color: #fff;
  border-radius: 5px;
  padding: 10px;
  font-weight: 600;
  justify-content: center;
  align-items: center;
  outline: none;
  transition: color 400ms, background-color 400ms;

  &:hover {
    background: #242AC6;
  }
`;
