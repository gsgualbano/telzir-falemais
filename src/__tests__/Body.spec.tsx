import React from 'react';
import { act } from 'react-dom/test-utils';
import Body from '../components/Body';

import { shallow } from '../enzyme';

describe('<Body />', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<Body />);
    expect(wrapper).toMatchSnapshot();
  });

  it('shows correct data', () => {
    const wrapper = shallow(<Body />);
    expect(wrapper.find('#origem').length).toEqual(1);
    expect(wrapper.contains('DDD Origem')).toBeTruthy();
    expect(wrapper.find('#destino').length).toEqual(1);
    expect(wrapper.contains('DDD Destino')).toBeTruthy();
    expect(wrapper.find('#tempo').length).toEqual(1);
    expect(wrapper.contains('Tempo de ligação(em minutos)')).toBeTruthy();
    expect(wrapper.find('#plano').length).toEqual(1);
    expect(wrapper.contains('Plano')).toBeTruthy();
  });
});
