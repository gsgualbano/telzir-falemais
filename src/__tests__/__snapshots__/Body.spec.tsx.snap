// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Body /> renders correctly 1`] = `
<styled.div>
  <styled.div>
    <styled.p>
      Com o nosso novo plano FaleMais, o cliente pode falar de graça até um determinado tempo e depois só paga os minutos excedentes. calcule e veja as vantagens
    </styled.p>
  </styled.div>
  <styled.div>
    <styled.div>
      <Animated
        animateOnMount={true}
        animationIn="fadeIn"
        animationInDelay={0}
        animationInDuration={1000}
        animationOut="fadeOut"
        animationOutDelay={0}
        animationOutDuration={1000}
        className=""
        isVisible={true}
        style={Object {}}
      >
        <Form
          onSubmit={[Function]}
          schema={
            ObjectSchema {
              "_blacklist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "_conditions": Array [],
              "_defaultDefault": [Function],
              "_deps": Array [],
              "_excludedEdges": Array [],
              "_exclusive": Object {},
              "_mutate": undefined,
              "_nodes": Array [
                "plano",
                "tempo",
                "destino",
                "origem",
              ],
              "_options": Object {
                "abortEarly": true,
                "recursive": true,
              },
              "_type": "object",
              "_typeError": [Function],
              "_whitelist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "fields": Object {
                "destino": StringSchema {
                  "_blacklist": RefSet {
                    "list": Set {},
                    "refs": Map {},
                  },
                  "_conditions": Array [],
                  "_deps": Array [],
                  "_exclusive": Object {
                    "required": true,
                  },
                  "_mutate": undefined,
                  "_options": Object {
                    "abortEarly": true,
                    "recursive": true,
                  },
                  "_type": "string",
                  "_typeError": [Function],
                  "_whitelist": RefSet {
                    "list": Set {},
                    "refs": Map {},
                  },
                  "tests": Array [
                    [Function],
                  ],
                  "transforms": Array [
                    [Function],
                  ],
                },
                "origem": StringSchema {
                  "_blacklist": RefSet {
                    "list": Set {},
                    "refs": Map {},
                  },
                  "_conditions": Array [],
                  "_deps": Array [],
                  "_exclusive": Object {
                    "required": true,
                  },
                  "_mutate": undefined,
                  "_options": Object {
                    "abortEarly": true,
                    "recursive": true,
                  },
                  "_type": "string",
                  "_typeError": [Function],
                  "_whitelist": RefSet {
                    "list": Set {},
                    "refs": Map {},
                  },
                  "tests": Array [
                    [Function],
                  ],
                  "transforms": Array [
                    [Function],
                  ],
                },
                "plano": StringSchema {
                  "_blacklist": RefSet {
                    "list": Set {},
                    "refs": Map {},
                  },
                  "_conditions": Array [],
                  "_deps": Array [],
                  "_exclusive": Object {
                    "required": true,
                  },
                  "_mutate": undefined,
                  "_options": Object {
                    "abortEarly": true,
                    "recursive": true,
                  },
                  "_type": "string",
                  "_typeError": [Function],
                  "_whitelist": RefSet {
                    "list": Set {},
                    "refs": Map {},
                  },
                  "tests": Array [
                    [Function],
                  ],
                  "transforms": Array [
                    [Function],
                  ],
                },
                "tempo": NumberSchema {
                  "_blacklist": RefSet {
                    "list": Set {},
                    "refs": Map {},
                  },
                  "_conditions": Array [],
                  "_deps": Array [],
                  "_exclusive": Object {
                    "required": true,
                  },
                  "_mutate": undefined,
                  "_options": Object {
                    "abortEarly": true,
                    "recursive": true,
                  },
                  "_type": "number",
                  "_typeError": [Function],
                  "_whitelist": RefSet {
                    "list": Set {},
                    "refs": Map {},
                  },
                  "tests": Array [
                    [Function],
                  ],
                  "transforms": Array [
                    [Function],
                  ],
                },
              },
              "tests": Array [],
              "transforms": Array [
                [Function],
              ],
            }
          }
        >
          <styled.div>
            <styled.label
              htmlFor="origem"
            >
              DDD Origem
            </styled.label>
            <Styled(Select)
              id="origem"
              name="origem"
              options={
                Array [
                  Object {
                    "id": "11",
                    "title": "11",
                  },
                  Object {
                    "id": "16",
                    "title": "16",
                  },
                  Object {
                    "id": "17",
                    "title": "17",
                  },
                  Object {
                    "id": "18",
                    "title": "18",
                  },
                ]
              }
            />
          </styled.div>
          <styled.div>
            <styled.label
              htmlFor="destino"
            >
              DDD Destino
            </styled.label>
            <Styled(Select)
              id="destino"
              name="destino"
              options={
                Array [
                  Object {
                    "id": "11",
                    "title": "11",
                  },
                  Object {
                    "id": "16",
                    "title": "16",
                  },
                  Object {
                    "id": "17",
                    "title": "17",
                  },
                  Object {
                    "id": "18",
                    "title": "18",
                  },
                ]
              }
            />
          </styled.div>
          <styled.div>
            <styled.label
              htmlFor="tempo"
            >
              Tempo de ligação(em minutos)
            </styled.label>
            <Styled(Input)
              id="tempo"
              name="tempo"
              placeholder=""
            />
          </styled.div>
          <styled.div>
            <styled.label
              htmlFor="plano"
            >
              Plano
            </styled.label>
            <Styled(Select)
              id="plano"
              name="plano"
              options={
                Array [
                  Object {
                    "id": "FaleMais 30",
                    "title": "FaleMais 30",
                  },
                  Object {
                    "id": "FaleMais 60",
                    "title": "FaleMais 60",
                  },
                  Object {
                    "id": "FaleMais 120",
                    "title": "FaleMais 120",
                  },
                ]
              }
            />
          </styled.div>
          <styled.button
            data-testid="form-button"
          >
            Calcular
          </styled.button>
        </Form>
      </Animated>
    </styled.div>
  </styled.div>
</styled.div>
`;
